### PR TITLE
test: add Hypothesis fuzz coverage and macos-latest CI matrix

### LINKED ISSUE #71: claude-code-chat-browser: Add JSONL parser fuzz tests and macos-latest CI matrix
## Calendar Day

Wednesday, June 11, 2026

## Planned Effort

**8 story points** (one PR) — combines sprint items **#4 + #6**:

| Sprint item | Points | Topic |
|-------------|--------|--------|
| #4 | 5 | Fuzz/adversarial testing for JSONL parser |
| #6 | 3 | macOS CI in matrix |

**Depends on:**

- Monday #1 — ruff + pip-audit job must run on `macos-latest` once matrix expands
- Tuesday #3 — narrowed TypedDicts define fuzz oracle (merge TypedDict PR before fuzz PR)

**Out of scope:** performance benchmarks (Thursday #7), frontend changes, parser API redesign.

## Problem

1. **Verification gap:** ~4,500 lines of tests cover known-good JSONL and fixtures, but no fuzz or adversarial testing targets `utils/jsonl_parser.py`. Schema drift from upstream Claude Code is discovered via user reports, not CI.
2. **Portability gap:** CI runs on `ubuntu-latest` and `windows-latest` (Week 1) but not `macos-latest`, despite macOS setup docs and POSIX `fcntl` locking that differs from Windows `msvcrt`.

## Goal

One merged PR that:

1. Adds **Hypothesis** fuzz tests proving `parse_session` does not crash on adversarial inputs.
2. Extends the CI matrix to **`macos-latest`** for all cross-platform test jobs (and Monday’s lint/audit job).
3. Keeps fuzz CI runtime **under ~60 seconds** via bounded example counts.

### A — Fuzz / adversarial tests (5 pt)

**Touch points:** `requirements-dev.txt`, `tests/test_parser_fuzz.py` (new), optional `tests/conftest.py` / `pyproject.toml` Hypothesis profile, `utils/jsonl_parser.py` (fixes only if fuzz exposes bugs)

- Add `hypothesis>=6.100.0` to dev dependencies.
- Create `tests/test_parser_fuzz.py` targeting `parse_session(filepath)` via temp `.jsonl` files.

**Required strategy coverage:**

| Category | Approach |
|----------|----------|
| Malformed JSON lines | Random text lines, invalid JSON |
| Truncated lines | Partial JSON (concurrent-write simulation) |
| Unknown record types | `type` not in `user`/`assistant`/`system`/`progress` |
| Missing / extra fields | `st.fixed_dictionaries` with optional keys removed or added |
| Deep nesting | Recursive JSON dict/list strategies |
| Long lines | 10k–50k char payloads inside JSON |
| Empty lines | Blank lines between records |
| Null bytes | Binary-safe write + UTF-8 decode with `errors="replace"` |

**Invariants:**

- `parse_session` must **not raise unhandled exceptions** for any fuzzed file input.
- Acceptable: normal `SessionDict` return; skipped malformed lines (existing behavior).
- If `SessionValidationError` or similar structured errors appear, either harden processors or document as allowed — prefer hardening so fuzz passes cleanly.

**Required explicit adversarial test:**

- Unknown `type` value (e.g. `totally-new-claude-record`) plus one valid `user` line → session parses, `entry_counts` reflects unknown type, no crash.

**CI budget:**

- `@settings(max_examples=200)` per strategy (tune down if needed).
- CI profile via `conftest.py` when `CI=true` (e.g. `max_examples=100`).
- Full fuzz module should complete in **< 60s** on `ubuntu-latest`.

**Ground truth for structured strategies:** `tests/test_jsonl_parser.py`, parametrized tool fixtures, Tuesday TypedDict unions in `models/`.

### B — macOS CI matrix (3 pt)

**Touch points:** `.github/workflows/ci.yml`, `README.md` or `CONTRIBUTING.md`

Add `macos-latest` to `strategy.matrix.os` for:

| Job | macOS? |
|-----|--------|
| `pytest` | yes |
| `integration-tests` | yes |
| `js-tests` | yes |
| `lint-and-audit` (Monday) | yes |
| `mypy` | no (ubuntu-only) |
| `prod-install-smoke` | no (bash heredoc) |

**Requirements:**

- All existing tests pass on macOS without new platform skips (unless genuine difference documented).
- **`fcntl.flock`** export lock path exercised on macOS (existing `tests/test_export_state_store.py` should suffice).
- Monday **ruff** + **pip-audit** steps pass on macOS runner.
- macOS job time ≤ **2×** Ubuntu job (monitor first run; optimize example count if needed).

**Docs:** Note three-platform CI in README or CONTRIBUTING.

### Fuzz (#4)

- [ ] `tests/test_parser_fuzz.py` exists using Hypothesis
- [ ] Strategies cover all categories listed in scope
- [ ] Parser does not raise unhandled exceptions on fuzzed inputs
- [ ] At least one adversarial test for unknown `type` with graceful degradation
- [ ] Fuzz tests run in CI with bounded iteration count; total fuzz time < ~60s on Ubuntu
- [ ] `pytest -q` passes locally and in CI

### macOS CI (#6)

- [ ] `macos-latest` in CI matrix for pytest, integration, js-tests, lint-and-audit
- [ ] All tests pass on macOS without undocumented skips
- [ ] `fcntl` locking path verified on macOS CI
- [ ] ruff + pip-audit green on macOS
- [ ] README or CONTRIBUTING documents cross-platform CI

### General

- [ ] `mypy` and `ruff check .` pass
- [ ] PR approved by at least 1 reviewer

### PR DESCRIPTION
**Pull request by clean6378-max-it**
closes #71 

## Summary

- Add Hypothesis fuzz tests for `parse_session` (malformed/truncated JSONL, unknown types, adversarial fields)
- Harden `utils/jsonl_parser.py` so fuzzed inputs do not raise unhandled exceptions
- Extend CI matrix to `macos-latest` (pytest, integration, vitest, lint-and-audit)
- Document Ubuntu + Windows + macOS CI in CONTRIBUTING

## Sprint items

Closes Chen W2 #4, #6 (single PR)

## Test plan

- [ ] `pytest tests/test_parser_fuzz.py -v`
- [ ] `pytest -q`
- [ ] `mypy -p api -p utils -p models`
- [ ] `ruff check .`
- [ ] `npm test`
- [ ] GitHub Actions: `macos-latest` matrix legs green for pytest, integration-tests, js-tests, lint-and-audit

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved JSONL parser robustness to gracefully handle malformed data, including non-numeric token values, unknown record types, and invalid nested fields.
  * Enhanced timestamp parsing to strictly validate ISO format entries.

* **Tests**
  * Added comprehensive fuzz testing suite for parser validation across edge cases and adversarial inputs.

* **Chores**
  * Expanded CI test matrix to include macOS alongside Ubuntu and Windows.
  * Removed optional dependencies from build configuration.

### CODE REVIEW

**Comment by clean6378-max-it** [comment]
@codeRabbit. please review this PR in more depth

**Comment by clean6378-max-it** [comment]
CR #4 (sys.path anti-pattern): factually wrong — tests/test_jsonl_parser.py:8 does the identical sys.path.insert, so it's the existing project convention. Removing it risks import breakage.
CR #9 (Rollup exit 1 fallback): intentional fail-fast; every OS in the matrix is covered, so the branch is unreachable today and a hard failure is the correct signal if an uncovered runner is ever added.
CR #10 (three coverage artifacts): pre-existing per-OS artifact naming (coverage-report-${{ matrix.os }}), not introduced by this PR; no merge step is needed unless Codecov is added later.


@codeRabbit

File: utils/jsonl_parser.py [hunk 1]
```diff
@@ -40,6 +40,16 @@
 ]
 
 
+def _safe_int(val: Any) -> int:
+    """Coerce a value to int for token accounting; non-numeric input becomes 0
+    so fuzzed/malformed usage fields never raise during arithmetic."""
+    if isinstance(val, bool):
+        return 0
+    if isinstance(val, (int, float)):
+        return int(val)
```
Comment (timon0305) [inline utils/jsonl_parser.py]: This still crashes on NaN/Infinity. Python's json. loads accepts the lierals NaN, Infinity, -Infinity by default, so an adversarial line like {"type":"assistant","message":{"usage":{"input_tokens":NaN}}} parses fine, reaches here as a float, and int(float('nan')) raises ValueError (int(inf) raises OverflowError) — which propagates out of parse_session. That's exactly the "never raise on malformed input" invariant this PR is about.

File: tests/test_parser_fuzz.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+_RECORD_TYPES = st.sampled_from(
+    ["user", "assistant", "system", "progress", "totally-new-claude-record", "future-record-v99"]
+)
+
+_json_leaf = st.one_of(
+    st.none(),
+    st.booleans(),
+    st.integers(),
+    st.floats(allow_nan=False, allow_infinity=False),
```
Comment (timon0305) [inline tests/test_parser_fuzz.py]: Because NaN/inf are excluded here, the fuzzer can never generate the input that breaks _safe_int:48 — which is why the suite stays green despite that gap. Add one explicit case writing a raw line with the literal NaN/Infinity

File: pyproject.toml [hunk 3]
```diff
@@ -27,6 +27,9 @@ select = ["E", "F", "W", "I"]
 [tool.ruff.lint.isort]
 combine-as-imports = true
 
+[tool.hypothesis]
+max_examples = 200
```
Comment (timon0305) [inline pyproject.toml]: Hypothesis doesn't read settings from pyproject.toml; the real example counts come from the profiles in conftest.py:17–19 (dev=200 / ci=100). This block has no effect — remove it, otherwise someone tunes this number expecting it to matter.

File: utils/jsonl_parser.py [hunk 4]
```diff
@@ -287,10 +302,10 @@ def _process_assistant(
...
             "usage": {
-                "input_tokens": usage.get("input_tokens") or 0,
-                "output_tokens": usage.get("output_tokens") or 0,
-                "cache_read": usage.get("cache_read_input_tokens") or 0,
-                "cache_creation": usage.get("cache_creation_input_tokens") or 0,
+                "input_tokens": _safe_int(usage.get("input_tokens")),
+                "output_tokens": _safe_int(usage.get("output_tokens")),
+                "cache_read": _safe_int(usage.get("cache_read_input_tokens")),
+                "cache_creation": _safe_int(usage.get("cache_creation_input_tokens")),
                 "service_tier": usage.get("service_tier"),
```
Comment (timon0305) [inline utils/jsonl_parser.py]: Line 309 stores the raw value while line 253 correctly guards it with isinstance(tier, str). Harmless (it's only echoed, never used in arithmetic), but a non-string tier leaks into the message payload — wrap it the same way for consistency.

File: utils/jsonl_parser.py [hunk 5]
```diff
@@ -109,11 +122,10 @@ def parse_session(filepath: str) -> SessionDict:
...
 
-            # Count entry types
-            if entry_type:
-                metadata["entry_counts"][entry_type] = (
-                    metadata["entry_counts"].get(entry_type, 0) + 1
-                )
+            # Count entry types (upstream may send non-str discriminants)
+            if entry_type is not None:
+                type_key = entry_type if isinstance(entry_type, str) else str(entry_type)
+                metadata["entry_counts"][type_key] = metadata["entry_counts"].get(type_key, 0) + 1
```
Comment (timon0305) [inline utils/jsonl_parser.py]: Good fix for non-string type values. Side effect on line 126: switching from truthiness to is not None means {"type": ""} now creates an empty-string count key (previously skipped). Almost certainly fine — just confirming it's intentional.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -188,4 +188,19 @@ jobs:
...
+          ROLLUP_VERSION=$(node -p "require('./node_modules/rollup/package.json').version")
+          case "$(node -p "process.platform + '-' + process.arch")" in
+            darwin-arm64) PKG="@rollup/rollup-darwin-arm64" ;;
+            darwin-x64)   PKG="@rollup/rollup-darwin-x64" ;;
+            linux-x64)    PKG="@rollup/rollup-linux-x64-gnu" ;;
+            win32-x64)    PKG="@rollup/rollup-win32-x64-msvc" ;;
+            win32-arm64)  PKG="@rollup/rollup-win32-arm64-msvc" ;;
+            *) echo "Unsupported Node platform: $(node -p "process.platform + '-' + process.arch")"; exit 1 ;;
+          esac
+          npm install --no-save "${PKG}@${ROLLUP_VERSION}"
```
Comment (timon0305) [inline .github/workflows/ci.yml:205]: Correct workaround for the npm optional-deps bug, and the *) … exit 1 on line 203 is the right fail-loud default. Nit: this step lives only in js-tests (starts line 171). If any other matrix leg later runs npm ci + a Rollup-backed build, it'll need the same step — worth factoring into a composite action to avoid drift.

File: package.json
Comment (timon0305) [inline package.json:1]: Dropping the linux-only optionalDependencies: "@rollup/rollup-linux-x64-gnu" pin is the right call — that hardcoded binary is what made macOS/Windows installs fragile; the CI step now resolves the right per-platform package instead. 👍

File: tests/conftest.py [hunk 7]
```diff
@@ -2,14 +2,22 @@
...
 import pytest
+from hypothesis import settings
 
 from app import create_app
 
+# Hypothesis profiles drive fuzz example counts/deadlines (deadline disabled to
+# avoid timing flakiness on slow/CI runners). CI runs fewer examples for speed.
+settings.register_profile("dev", max_examples=200, deadline=None)
+settings.register_profile("ci", max_examples=100, deadline=None)
+settings.load_profile("ci" if os.environ.get("CI") else "dev")
```
Comment (timon0305) [inline tests/conftest.py:19]: Clean profile setup. Only note: deadline=None disables per-example timing suite-wide — sensible for fuzz on slow runners, just means a genuinely slow parser regression won't trip a deadline. Fine to keep.

**Review by timon0305 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 6 above)
Comment (clean6378-max-it) [reply .github/workflows/ci.yml:205]: Rollup step only in js-tests / composite action. Legitimate future-proofing nit, but js-tests is the only leg that runs npm ci today, so there's no drift yet. Keeping changes minimal — skipped.

**Comment by clean6378-max-it** [comment]
@codeRabbit. do you think my last commits since @timon0305 's comments  addressed all?

Currently they are not resolved yet so I am asking you to check thosee again and if they are addressed ,I will resolve t hose off clearly.

File: pyproject.toml (hunk 3 above)
Comment (clean6378-max-it) [reply pyproject.toml]: Comment already documents forward-looking intent; no behavior change needed

**Review by timon0305 — APPROVED** [review]

**Review by wpak-ai — APPROVED** [review]
